### PR TITLE
Create a System Builder Syntax instead of using System Descriptors

### LINKED ISSUE #1978: Extensible builder syntax for system insertion
## What problem does this solve or what need does it fill?

As we create more and more interesting ways to initialize systems, the number of AppBuilder methods we require explodes in an exponential fashion.

`add_system` becomes `add_startup_system` becomes `add_startup_system_to_stage` become `add_startup_system_set_to_stage` becomes...

## What solution would you like?

Use two base methods `add_system` and `add_system_set`. Then, use the builder pattern (see `EntityCommands` for an example) to replicate the existing API using method chaining.

For example, `add_startup_system_set_to_stage(my_stage, my_system_set)` would become `add_system_set(my_system_set).startup().to_stage(my_stage)`.

## What alternative(s) have you considered?

Continue expanding our API and method names. That's what IDEs are for right? ;)

## Additional context

This would help enable the [subworlds RFC](https://github.com/bevyengine/rfcs/pull/16) and similar extensions.

This is a very nice building block for "labels of mass description", hinted at in #1375. It also opens the door to principled modification of systems added by plugins (e.g. to move them into the required States).

### PR DESCRIPTION
**Pull request by thechubbypanda**
# Objective

Closes #1978

## Solution

Working around @cart's [design](https://github.com/bevyengine/bevy/issues/1978#issuecomment-825337056], implement a builder syntax for systems such that we can do something like:
```
App::new().add_system(
    foo.add_to_stage(Stage::Label)
        .before(Stage::Label2)
        .startup()
)
```

Please do watch this pr and comment on the changes made :)
Made with love and significant help from @Ratysz XD

### CODE REVIEW

File: crates/bevy_ecs/src/schedule/run_criteria.rs [hunk 1]
```diff
@@ -457,11 +457,6 @@ impl System for RunOnce {
 
     fn check_change_tick(&mut self, _change_tick: u32) {}
 
-    fn run(&mut self, input: Self::In, world: &mut World) -> Self::Out {
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/schedule/run_criteria.rs]: Out of curiosity, why is this removed?

**Comment by alice-i-cecile** [comment]
Documentation checklist:
- [x] all doc examples and ordinary examples compile (should be done by CI already)
- [ ] ensure that `System` docs point to these configuration methods: this will be the most obvious thing to search
- [ ] point to the main docs for each configuration method (stages, states, labels, system ordering, exclusive systems)
- [ ] point to the configuration methods for each relevant concept 
- [ ] update the book to reflect these changes (this one is on me and other authors over in the `bevy-website` repo once this is merged)

**Comment by thechubbypanda** [comment]
Alice, could you please clarify what you mean by
> main docs for each configuration method
and also
> point to the configuration methods for each relevant concept
from where?

File: crates/bevy_ecs/src/schedule/stage.rs [hunk 2]
```diff
@@ -49,7 +49,7 @@ impl_downcast!(Stage);
 pub struct ReportExecutionOrderAmbiguities;
 
 /// Stores and executes systems. Execution order is not defined unless explicitly specified;
-/// see `SystemDescriptor` documentation.
+/// see `SystemConfig` documentation.
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/schedule/stage.rs:52]:
```suggestion
/// see `SystemConfig` documentation to configure which stage a system runs in.
```

File: crates/bevy_ecs/src/system/config/system_config.rs [hunk 3]
```diff
@@ -1,5 +1,6 @@
 use crate::{schedule::*, system::InsertionPoint};
 
+/// Each system has one of these. It is updated using the various traits seen in this [directory](super).
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/system/config/system_config.rs:3]:
```suggestion
/// Controls when and how a [`System`] is run
///
/// The traits in this [module](super) are used to update these settings.
```

Not thrilled with this, but it's an improvement

File: crates/bevy_ecs/src/system/system.rs [hunk 4]
```diff
@@ -31,7 +31,7 @@ impl SystemId {
 ///
 /// Systems are executed in parallel, in opportunistic order; data access is managed automatically.
 /// It's possible to specify explicit execution order between specific systems,
-/// see [SystemDescriptor](crate::schedule::SystemDescriptor).
+/// see [SystemConfig](super::SystemConfig) and the other files in the [./config](super::config) directory.
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/system/system.rs:34]:
```suggestion
/// see [SystemConfig](super::SystemConfig) and the other members of the [`config` module](super::config).
```

Users won't / shouldn't have to dig through the source, so "directory" is inappropriate.

File: crates/bevy_ecs/src/schedule/run_criteria.rs (hunk 1 above)
Comment (thechubbypanda) [reply crates/bevy_ecs/src/schedule/run_criteria.rs]: It's predefined as a default, just extra lines if it's included :man_shrugging:

File: crates/bevy_ecs/src/system/config/system_config.rs (hunk 3 above)
Comment (thechubbypanda) [reply crates/bevy_ecs/src/system/config/system_config.rs]: My grammar isn't amazing, sorry

**Comment by alice-i-cecile** [comment]
> main docs for each configuration method

So, this is "the method to add a system to a stage should point to the Stage docs" and so on.

>> point to the configuration methods for each relevant concept

> from where?

From e.g. the Stage docs. The idea here is that we want bidirectional links between "the thing we're configuring" and "the tool used to configure it".

**Comment by thechubbypanda** [comment]
> Documentation checklist

I believe all are now complete bar the book :)

**Comment by IceSentry** [comment]
Could this also be ported to `SystemSet`s? That would allow things like this:

```rust
.add_system_set(
	SystemSet::new().stage(Stage::Label)
		.with_system(example_system))
```

instead of the current approach of

```rust
.add_system_set_to_stage(
	Stage::Label, 
	SystemSet::new()
		.with_system(example_system))
```

I'm also curious what happens if a system in a SystemSet specifies a stage like so

```rust
.add_system_set_to_stage(
	Stage::Label, 
	SystemSet::new()
		.with_system(example_system.stage(Stage::OtherLabel)))
```

**Comment by Ratysz** [comment]
`.add_system_set_to_stage()` shouldn't be a thing anymore. If it is, that's an oversight in the PR.

**Comment by IceSentry** [comment]
@Ratysz yeah, sorry that was copy pasted from the first part of my question before I noticed it was removed and I removed that part of my comment. It's not in the code anymore.

**Comment by Ratysz** [comment]
Right. I don't remember if it came up during implementation, but that should crash, the same as with extra run criteria.

**Comment by thechubbypanda** [comment]
We don't actually throw any errors when a system has it's own stage and is also inside a set. On baking the SystemSet, we either overwrite or add to data in the systems.

**Comment by Ratysz** [comment]
Well... I think both the system and the set it's in should be a hard error, because it's code that doesn't have an obvious correct pathway. See [here](https://github.com/bevyengine/bevy/pull/1675#discussion_r599150245) for similar discussion about run criteria.

File: crates/bevy_app/src/app.rs [hunk 5]
```diff
@@ -203,85 +201,89 @@ impl App {
...
-    pub fn add_startup_system_to_stage<Params>(
-        &mut self,
-        stage_label: impl StageLabel,
-        system: impl IntoSystemDescriptor<Params>,
-    ) -> &mut Self {
-        self.schedule
-            .stage(CoreStage::Startup, |schedule: &mut Schedule| {
-                schedule.add_system_to_stage(stage_label, system)
-            });
+    pub fn add_system_set(&mut self, set: SystemSet) -> &mut Self {
```
Comment (IceSentry) [inline crates/bevy_app/src/app.rs:231]: I think there should be some kind of validation here if a system inside the system set has a conflicting stage.

**Review by IceSentry — APPROVED** [review]
It's not particularly important to fix this, but some of the impl block have inconsistent spacing between functions. MOst of the codebase has a space between each function.

Overall, I really like this new api.

File: crates/bevy_app/src/app.rs [hunk 6]
```diff
@@ -203,85 +201,89 @@ impl App {
...
-        &mut self,
-        stage_label: impl StageLabel,
-        system_set: SystemSet,
-    ) -> &mut Self {
-        self.schedule
-            .add_system_set_to_stage(stage_label, system_set);
+    pub fn add_system<Params>(&mut self, system: impl IntoSystem<(), (), Params>) -> &mut Self {
+        let system = system.system();
+        if system.config().startup {
+            let system = {
```
[the remark this answers is not in the document]
Comment (thechubbypanda) [reply crates/bevy_app/src/app.rs:207]: I think this is required because we need to access config and unless we call the .system() we can't actually access it.

[the remark this answers is not in the document]
Comment (IceSentry) [reply crates/bevy_app/src/app.rs:207]: I just mean the block for 
```rust
let system = {
	if system.config().stage.is_some() {
    	system
  	} else {
      	system.stage(StartupStage::Startup)
	}
}
//vs
let system = if system.config().stage.is_some() {
	system
} else {
	system.stage(StartupStage::Startup)
};
```
I tried it locally and it seems to work fine

[the remark this answers is not in the document]
Comment (IceSentry) [reply crates/bevy_app/src/app.rs:207]: @thechubbypanda I'm pinging you because I don't think you'll see this otherwise since it's resolved. 

Also, this isn't a required change or anything I was just curious if there was a reason for this or if it's just a style preference.

File: crates/bevy_app/src/app.rs [hunk 7]
```diff
@@ -229,6 +229,23 @@ impl App {
     }
 
     pub fn add_system_set(&mut self, set: SystemSet) -> &mut Self {
+        fn ensure_no_conflicts(set: &SystemSet) {
```
Comment (IceSentry) [inline crates/bevy_app/src/app.rs:232]: I think this could keep a reference to the conflicting stage name and add that to the panic message.

Small suggestion, you could use this instead `conflict = set.systems.iter().any(|system| system.config().stage != set.config().stage);`, but that's entirely up to preference.

**Comment by alice-i-cecile** [comment]
Fixed by #45.
